include/quic/datagram.hpp

```diff
@@ -0,0 +1,95 @@
+    using io_unblocked_callback = std::function<bool(IOChannel&)>;
+
+    class IOChannel
+    {
+      public:
+        IOChannel(Connection& c, Endpoint& e, io_data_callback data_cb = nullptr, io_close_callback close_cb = nullptr);
+
+        virtual ~IOChannel() { log::trace(log_cat, "{} called", __PRETTY_FUNCTION__); };
+
+        Connection& conn;
+        Endpoint& endpoint;
+        io_data_callback data_callback;
+        io_close_callback close_callback;
+        buffer_que user_buffers;
+
+        // no copy, no move. always hold in a shared pointer
+        IOChannel(const IOChannel&) = delete;
+        IOChannel& operator=(const IOChannel&) = delete;
+        IOChannel(IOChannel&&) = delete;
+        IOChannel& operator=(IOChannel&&) = delete;
+
+        virtual bool is_stream() = 0;
+        virtual bool is_empty() const = 0;
+        virtual std::shared_ptr<Stream> get_stream() = 0;
+        virtual void send(bstring_view data, std::shared_ptr<void> keep_alive = nullptr) = 0;
+        virtual std::vector<ngtcp2_vec> pending() = 0;
+        virtual void append_buffer(bstring_view buffer, std::shared_ptr<void> keep_alive) = 0;
+        virtual int64_t stream_id() const = 0;
+        virtual bool is_closing() const = 0;
+        virtual bool sent_fin() const = 0;
+        virtual void set_fin(bool v) = 0;
+        virtual size_t unsent() const = 0;
+        virtual void wrote(size_t n) = 0;
+    };
+
+    class DatagramIO : public IOChannel
+    {
+      public:
+        DatagramIO(Connection& c, Endpoint& e, io_data_callback data_cb = nullptr, io_close_callback close_cb = nullptr);
+
+        bool is_empty() const override { return user_buffers.empty(); }
+
+        void send(bstring_view data, std::shared_ptr<void> keep_alive = nullptr) override;
+        void append_buffer(bstring_view buffer, std::shared_ptr<void> keep_alive) override;
+        std::vector<ngtcp2_vec> pending() override;
+
+        bool is_stream() override { return false; }
+
+      private:
+        std::shared_ptr<Stream> get_stream() override
+        {
+            log::debug(log_cat, "{} called", __PRETTY_FUNCTION__);
+            return nullptr;
+        };
+        int64_t stream_id() const override
+        {
+            log::debug(log_cat, "{} called", __PRETTY_FUNCTION__);
+            return std::numeric_limits<int64_t>::min();
+        };
+        bool is_closing() const override
+        {
+            log::debug(log_cat, "{} called", __PRETTY_FUNCTION__);
+            return false;
+        };
+        bool sent_fin() const override
+        {
+            log::debug(log_cat, "{} called", __PRETTY_FUNCTION__);
+            return false;
+        };
+        void set_fin(bool) override { log::debug(log_cat, "{} called", __PRETTY_FUNCTION__); };
+        size_t unsent() const override
+        {
+            log::debug(log_cat, "{} called", __PRETTY_FUNCTION__);
+            return std::numeric_limits<size_t>::min();
```

Comment:
Feels like this should have an actual value it could return.

Reply:
For sure, I just needed a placeholder here in case it got accidentally called that wasn't in the realm of a realistic response